### PR TITLE
[draft, DNM] mass conversion of k_work API

### PR DESCRIPTION
**Pull request by pabigot**
Use coccinelle to do a blanket translation from the deprecated API to the replacement API.

Conversions have been grouped into areas for convenience. There is no guarantee of bisectability between the resulting commits, as conversions in one commit may depend on conversions in another commit. Conversions of remaining time API are not optimized.

NOTE: These commits are not tested and not intended to be merged as is or via this PR.  Area maintainers can cherry-pick or split the commits to create new PRs that change the API and address any relevant race conditions or other problems.  See zephyrproject-rtos/zephyr#33104 and the documentation update in zephyrproject-rtos/zephyr#34466 for details on the API change and best practices to avoid mis-use of the work API.

### CODE REVIEW

File: applications/nrf_desktop/src/modules/ble_scan.c [hunk 1]
```diff
@@ -424,7 +424,7 @@ static void scan_connecting_error(struct bt_scan_device_info *device_info)
 {
 	LOG_WRN("Connecting failed");
 	scan_counter = SCAN_TRIG_TIMEOUT_MS;
-	k_delayed_work_submit(&scan_start_trigger,
```
Comment (pdunaj) [inline applications/nrf_desktop/src/modules/ble_scan.c:427]: Ok. There is a potential race with scan_start and its cancel but I think it is impossible to occur as this callback will trigger when scanning was in progress and connection is attempted.

File: applications/nrf_desktop/src/modules/dfu.c [hunk 2]
```diff
@@ -148,8 +148,8 @@ static void terminate_dfu(void)
 	__ASSERT_NO_MSG(flash_area != NULL);
 
 	flash_area_close(flash_area);
-	k_delayed_work_cancel(&dfu_timeout);
```
Comment (pdunaj) [inline applications/nrf_desktop/src/modules/dfu.c:151]: Ok. Cancels will not race with any resubmit - all operations are in sync with system workqueue.

File: applications/nrf_desktop/src/modules/hid_forward.c [hunk 3]
```diff
@@ -880,7 +880,7 @@ static void disconnect_peripheral(struct hids_peripheral *per)
 	__ASSERT_NO_MSG(!is_any_report_enqueued(&per->enqueued_reports));
 
 	bt_hogp_release(&per->hogp);
-	k_delayed_work_cancel(&per->read_rsp);
```
Comment (pdunaj) [inline applications/nrf_desktop/src/modules/hid_forward.c:883]: There is a potential race that will need to be carefully reviewed. The problem will happen if cancel will block resubmission done in the other thread. We need to check how these callbacks are called.

File: applications/nrf_desktop/src/modules/hids.c [hunk 4]
```diff
@@ -496,15 +496,15 @@ static void notify_hids(const struct ble_peer_event *event)
 		cur_conn = NULL;
 		secured = false;
 		if (CONFIG_DESKTOP_HIDS_FIRST_REPORT_DELAY > 0) {
-			k_delayed_work_cancel(&notify_secured);
```
Comment (pdunaj) [inline applications/nrf_desktop/src/modules/hids.c:499]: Ok. Cancel will not race with any resubmit - all operations are in sync with system workqueue.

File: applications/nrf_desktop/src/modules/power_manager.c [hunk 5]
```diff
@@ -262,7 +262,7 @@ static bool event_handler(const struct event_header *eh)
 		switch (event->state) {
 		case USB_STATE_POWERED:
 		case USB_STATE_ACTIVE:
-			k_delayed_work_cancel(&power_down_trigger);
```
Comment (pdunaj) [inline applications/nrf_desktop/src/modules/power_manager.c:265]: Ok. Cancels will not race with any resubmit - all operations are in sync with system workqueue.

File: applications/nrf_desktop/src/util/config_channel_transport.c [hunk 6]
```diff
@@ -262,7 +262,7 @@ bool config_channel_transport_rsp_receive(struct config_channel_transport *trans
 	ARG_UNUSED(pos);
 
 	transport->state = CONFIG_CHANNEL_TRANSPORT_RSP_READY;
-	k_delayed_work_cancel(&transport->timeout);
```
Comment (pdunaj) [inline applications/nrf_desktop/src/util/config_channel_transport.c:265]: Will have to revisit it as dependencies here are more complex.

**Review by pdunaj — COMMENTED** [review]
nrf_desktop part

**Comment by trond-snekvik** [comment]
@pdunaj Thanks for the feedback. For this to be finished in time, I think the individual code owners should implement the required changes, preferably in separate PRs.

**Comment by pdunaj** [comment]
> @pdunaj Thanks for the feedback. For this to be finished in time, I think the individual code owners should implement the required changes, preferably in separate PRs.

I imagine. I have noted what I suspect may be wrong. Will go through the caf in a moment. If there is anything to fix we will fix it.

**Comment by pdunaj** [comment]
If anybody is interested.
The new implementation fixes some important bugs. There is however one big difference that needs to be assessed. The old cancel simply removed the work's node from the list. It would mean that when you canceled you could right away do resubmission. In the new implementation that is not possible - if the work note was picked for run, the work's handler must finish and cancelation flag must be cleared.
It means that when you cancel/resubmit from various contexts you may end up with your resubmit being rejected because cancel flag was not cleared. The good thing is that it returns an error so if there are any doubts it can be asserted and tested for races.

**Pull request closed without merging**